PerimeterXModule/Internals/ReverseProxy.cs

```diff
@@ -0,0 +1,240 @@
+﻿using PerimeterX;
+using System;
+using System.Collections.Specialized;
+using System.Diagnostics;
+using System.Net;
+using System.Text;
+using System.Web;
+
+namespace PerimeterX
+{
+	public interface IReverseProxy
+	{
+		void ReversePxClient(HttpContext context);
+		void ReversePxXhr(HttpContext context);
+		bool ShouldReverseClient(string uri);
+		bool ShouldReverseXhr(string uri);
+	}
+
+	public class ReverseProxy : IReverseProxy
+	{
+		private readonly string DEFAULT_CLIENT_VALUE = "";
+		private readonly string CONTENT_TYPE_JAVASCRIPT = "application/javascript";
+		private readonly string DEFAULT_JSON_VALUE = "{}";
+		private readonly string CONTENT_TYPE_JSON = "application/json";
+		private readonly byte[] DEFAULT_EMPTY_GIF_VALUE = { 0x47 , 0x49, 0x46, 0x38, 0x39, 0x61, 0x01, 0x00, 0x01, 0x00, 0x80, 0x00,
+		0x00, 0xff, 0xff, 0xff, 0x00, 0x00, 0x00, 0x2c, 0x00, 0x00, 0x00, 0x00,
+		0x01, 0x00, 0x01, 0x00, 0x00, 0x02, 0x02, 0x44, 0x01, 0x00, 0x3b };
+		private readonly string CONTENT_TYPE_GIF = "image/gif";
+
+		private readonly string XHR_PATH = "xhr";
+		private readonly string CLIENT_PATH = "init.js";
+
+		private string ClientReversePrefix;
+		private string XhrReversePrefix;
+		private string CollectorUrl;
+
+		public bool IsReusable
+		{
+			get { return true; }
+		}
+
+		private PxModuleConfigurationSection PxConfig { get; set; }
+
+		public ReverseProxy(PxModuleConfigurationSection pxConfig)
+		{
+			PxConfig = pxConfig;
+			string prefixFormat = "/{0}/{1}";
+			string appIdPrefix = pxConfig.AppId.Substring(2);
+			ClientReversePrefix = string.Format(prefixFormat, appIdPrefix, CLIENT_PATH);
+			XhrReversePrefix = string.Format(prefixFormat, appIdPrefix, XHR_PATH);
+			CollectorUrl = string.Format(pxConfig.CollectorUrl, pxConfig.AppId);
+		}
+
+		/**
+		 *  <summary>
+		 * Method calls when client request the server
+		 * </summary>
+		 * <param name="context">HttpContext</param>
+		 * <param name="serverUrl">string</param>
+		 * <param name="uri">string</param>
+		 * <returns>bool, true if re</returns>
+		 */
+		private bool ProcessRequest(HttpContext context, string serverUrl, string uri)
+		{
+			try
+			{
+				// Set headers
+				context.Request.Headers.Add(PxConstants.ENFORCER_TRUE_IP_HEADER, PxCommonUtils.GetRequestIP(context, PxConfig));
+				context.Request.Headers.Add(PxConstants.FIRST_PARTY_HEADER, PxConstants.FIRST_PARTY_VALUE);
+
+				// Create a connexion to the Remote Server to redirect all requests
+				RemoteServer server = new RemoteServer(context, serverUrl, uri);
+
+				// Create a request with same data in navigator request
+				HttpWebRequest request = server.GetRequest();
+
+				// Send the request to the remote server and return the response
+				HttpWebResponse response = server.GetResponse(request);
```

Comment:
Since you don't really use the `request` (returned from `server.GetRequest()`) except for `server.GetResponse(request)`, why not generate the request _and_ send the response as part of the same operation?

Reply:
👎 
decoupling the logic
each method will serve its logic
and btw,

Reply:
I didn't mean to have a single function to do them both internally. The suggestion was to have one external function, and internally inside ReverseProxy.cs break it down to do `GetRequest()` and then `GetResponse()`. I see no reason to separate them on the consuming entity.

Reply:
👍